### PR TITLE
BUG/ENH: Improve output for structured non-void types

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes gh-9821

Based on the 1.14 branch point, in case we want to backport.

### CODE REVIEW

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -394,7 +391,12 @@ def _get_format_function(data, **options):
     elif issubclass(dtypeobj, _nt.object_):
         return formatdict['object']()
     elif issubclass(dtypeobj, _nt.void):
-        return formatdict['void']()
+        # StructureFormat relies on np.void.__getitem__, so we can't use it
```
Comment (mhvk) [inline numpy/core/arrayprint.py]: This comment would seem helpful only if you thought `StructureFormat` should be used elsewhere (as it was originally, but which is not all that obvious). Maybe just state here something along the lines of "'void' only deals with unstructured bytes; we use StructureFormat if they are assigned to fields. Note that we cannot use StructureFormat where the void is in a union with a non-void dtype; those will be caught above and the typeset with the regular dtype (though the full dtype will be printed, to indicate to the user it is a union dtype)"

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -1236,7 +1238,7 @@ def dtype_is_implied(dtype):
     dtype = np.dtype(dtype)
     if _format_options['legacy'] == '1.13' and dtype.type == bool_:
         return False
-    return dtype.type in _typelessdata
+    return dtype.type in _typelessdata and dtype.names is None
```
Comment (mhvk) [inline numpy/core/arrayprint.py]: As we're adding comments, maybe here note that the latter part guards against union dtypes?

File: numpy/core/tests/test_multiarray.py [hunk 3]
```diff
@@ -478,6 +478,24 @@ def test_byteorders(self):
...
 
+    def test_structured_non_void(self):
+        fields = [('a', '<i2'), ('b', '<i2')]
+
+        class CustomScalar(np.void):
+            pass
+        CustomScalar.__module__ = None
+
+        dt_custom = np.dtype((CustomScalar, fields))
+        str(dt_custom)  # segfault?
```
Comment (mhvk) [inline numpy/core/tests/test_multiarray.py]: Might as well test the output?

**Review by mhvk — COMMENTED** [review]
Looks good. Mostly nitpicks about comments, to help possible future developers.

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/arrayprint.py]: > If you thought StructureFormat should be used elsewhere
That depends on what you think counts as a structured type:

* `issubdtype(dt, np.void) and dt.names is not None`
* `dt.names is not None`
* `issubdtype(dt, np.flexible)` (a lot of old code refers to flexible types when it means structured types)

File: numpy/core/tests/test_multiarray.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/core/tests/test_multiarray.py]: Output is meaningless with `__module__` assigned to garbage, I think. Can `__module__` ever be `None` in real code?

File: numpy/core/src/multiarray/descriptor.c [hunk 4]
```diff
@@ -3356,42 +3356,37 @@ arraydescr_struct_str(PyArray_Descr *dtype, int includealignflag)
...
             return NULL;
         }
 
-        str_module = PyUString_FromString("__module__");
-        modulestr = PyObject_GetAttr((PyObject*)(dtype->typeobj), str_module);
-        Py_DECREF(str_module);
 
-        ret = PyUString_FromString("(");
-        if (modulestr != NULL) {
-            /* Note: if modulestr == NULL, the type is unpicklable */
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c]: More background behind this comment would be useful. Can this ever happen?

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (mhvk) [reply numpy/core/arrayprint.py]: Agreed it is confusing, but this is one place where one can explain what it is in the code!

File: numpy/core/tests/test_multiarray.py (hunk 3 above)
Comment (mhvk) [reply numpy/core/tests/test_multiarray.py]: Yes, true. I tried it myself and indeed got `__main__.CustomScalar` - I just wondered whether this is tested anywhere at all...

File: numpy/core/src/multiarray/descriptor.c (hunk 4 above)
Comment (ahaldane) [reply numpy/core/src/multiarray/descriptor.c]: Hmm my memory of this is vague now. I think this code had to do with "old style" vs "new style" classes in python2. I think "old style" classes fill in the "__module__" entry incompletely, or something along those lines. I will have to look it up.

Comment (ahaldane) [reply numpy/core/src/multiarray/descriptor.c]: Ah, happily I left some kind of explanation of that in a github comment. See [here (link)](https://github.com/numpy/numpy/pull/5483#issuecomment-71233980).

Now as to whether this can be NULL, I think it can, for types defined in C in python 2. If you go to https://docs.python.org/2/extending/newtypes.html, and copy the first "noddy" example, you will find:
```python
>>> import noddy
>>> mynoddy = noddy.Noddy()
>>> mynoddy.__module__
AttributeError: 'noddy.Noddy' object has no attribute '__module__'
>>> import cPickle
>>>  cPickle.dumps(mynoddy)
TypeError: can't pickle Noddy objects
```

**Comment by eric-wieser** [comment]
@mhvk: Comments updated

**Comment by charris** [comment]
I'll leave this to @mhvk and @ahaldane .

File: numpy/core/src/multiarray/descriptor.c (hunk 4 above)
Comment (mhvk) [reply numpy/core/src/multiarray/descriptor.c]: So, should the code continue to be able to deal with `__module__ == NULL`? Or is the new behaviour of raising an exception fine? It would probably be good to add a comment with the conclusion regardless...

Comment (ahaldane) [reply numpy/core/src/multiarray/descriptor.c]: It's probably fine to raise an exception if it is NULL, let's go with these changes.I would change the comment, though, it is technically wrong: Instead of ` this should never happen since types always have these attributes` it should say something like `we do not expect this to happen happen since types always have these attributes, except in some very special cases (module may be null for "statically allocated" types in python2)`.

Probably these "special cases" never come up in practice, since they will cause people other problems (eg, unpicklable types).

File: numpy/core/arrayprint.py [hunk 5]
```diff
@@ -1192,13 +1199,16 @@ def __call__(self, x):
         else:
             return "({})".format(", ".join(str_fields))
 
+# for backwards compatibility
+StructureFormat = StructuredVoidFormat
```
Comment (ahaldane) [inline numpy/core/arrayprint.py]: Should we add a deprecation warning for use of StructureFormat, the same way we did for LongFloatFormat?

Comment (eric-wieser) [reply numpy/core/arrayprint.py]: Sure, wouldn't hurt

File: numpy/core/src/multiarray/descriptor.c [hunk 6]
```diff
@@ -3356,42 +3356,37 @@ arraydescr_struct_str(PyArray_Descr *dtype, int includealignflag)
...
          */
 
-        PyObject *str_name, *namestr, *str_module, *modulestr, *ret;
+        PyObject *namestr, *modulestr, *fmt, *ret;
 
-        str_name = PyUString_FromString("__name__");
-        namestr = PyObject_GetAttr((PyObject*)(dtype->typeobj), str_name);
-        Py_DECREF(str_name);
+        namestr = PyObject_GetAttrString((PyObject *)dtype->typeobj, "__name__");
+        modulestr = PyObject_GetAttrString((PyObject *)dtype->typeobj, "__module__");
```
Comment (ahaldane) [inline numpy/core/src/multiarray/descriptor.c]: Some of these lines are too long, too

File: numpy/core/src/multiarray/descriptor.c (hunk 4 above)
Comment (ahaldane) [reply numpy/core/src/multiarray/descriptor.c]: One last comment: I looked more closely at this and I have not been able to construct an example where `__module__` is NULL.  The type `from numpy.core.test_rational import rational` is what I expected to fail, since it does not properly define a module, but if you do `rational.__module__` you get `__builtin__`, which is not NULL. I tested that `np.dtype((rational, 'i4,i4'))` works properly with this PR.

So in summary, this code is fine.

**Comment by ahaldane** [comment]
Everything else LGTM!

**Comment by charris** [comment]
Be good to finish this up.

**Comment by eric-wieser** [comment]
Will pick this up in the next few days

**Comment by eric-wieser** [comment]
I'm gonna roll back the segfault fix - we need a much wider audit of `PyUString_ConcatAndDel`, even within that file.

**Comment by eric-wieser** [comment]
Alright, that's the fixes for regular structured scalars at least

**Comment by ahaldane** [comment]
LGTM. Circleci fails, but I'll merge anyway.

Thanks Eric!
